src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java

```diff
@@ -216,6 +215,67 @@ private SqlSelectList parseSelectList(final JsonArray selectList) {
         }
     }
 
+    private List<SqlNode> collectAllInvolvedColumns(final SqlNode from) {
+        final List<SqlTable> involvedTables = collectInvolvedTables(from);
+        final Map<String, TableMetadata> tableMetadataMap = getInvolvedTablesMetadataMap();
+        final List<SqlNode> selectListElements = new ArrayList<>();
+        for (final SqlTable table : involvedTables) {
+            final String tableName = table.getName();
+            if (tableMetadataMap.containsKey(tableName)) {
+                final List<ColumnMetadata> columns = tableMetadataMap.get(tableName).getColumns();
+                for (int i = 0, columnsSize = columns.size(); i < columnsSize; ++i) {
+                    selectListElements.add(createColumn(i, table, columns.get(i)));
+                }
+            } else {
+                throw new IllegalStateException("Unable to find a metadata for table \"" + table.getName()
+                        + "\" during collection of involved columns.");
+            }
+        }
+        return selectListElements;
+    }
+
+    /**
+     * We collect all tables from the FROM clause in a given order. We need the tables, because the
+     * involvedTablesMetadata field doesn't contain `tableAlias` information.
+     */
+    private List<SqlTable> collectInvolvedTables(final SqlNode from) {
+        final List<SqlTable> involvedTables = new ArrayList<>();
+        final Queue<SqlNode> nodes = new LinkedList<>();
+        nodes.add(from);
+        while (!nodes.isEmpty()) {
+            final SqlNode node = nodes.poll();
+            switch (node.getType()) {
+            case TABLE:
+                involvedTables.add((SqlTable) node);
+                break;
+            case JOIN:
+                nodes.add(((SqlJoin) node).getLeft());
+                nodes.add(((SqlJoin) node).getRight());
+                break;
+            default:
+                throw new IllegalStateException(
+                        "Encountered illegal SqlNodeType during collection involved tables " + node.getType());
+            }
+        }
+        return involvedTables;
+    }
```

Comment:
The queue is confusing me. Shouldn't this function return the involved tables in the wrong order, if we have an input similar to the test `pushdown_request_nested_join_without_select_list.json`, but with the join on the left side and the table on the right side? Please add such a test. Somehow the select list in the generated SQL is correct in that case, but I would still prefer a recursive function here instead of a queue, to make this more readable.

Reply:
I'm going to test it